### ISSUE TITLE
multiple Application classes

### ISSUE THREAD

**Issue by tompasik** [mvc-spec/ozark#143]
Hey guys,

I've started playing around with MVC/Ozark and immediately stumbled upon an issue that may need resolving or at least discussing.

I've decided splitting template based resources (handled by MVC) from regular resources handled by JAX-RS implementation using different `javax.ws.rs.core.Application` classes annotated with `@ApplicationPath` in a way that _/views/*_ urls are handled by the former class and _/api/*_ by the latter.

Apparently such a scheme is not handled correctly by Ozark :-( The JAX-RS spec does not seem to constrain this (multiple classes extending `javax.ws.rs.core.Application` classes are not forbidden and hence allowed ? :-) ) and the MVC spec does not seem to mention anything regarding this - so I would assume this should be handled just fine.

I've been testing this on Payara-5 and got the following error:

```
  Error invoking ServletContainerInitializer org.mvcspec.ozark.servlet.OzarkContainerInitializer
java.lang.IllegalStateException: More than one JAX-RS ApplicationPath detected!
	at org.mvcspec.ozark.servlet.OzarkContainerInitializer.onStartup(OzarkContainerInitializer.java:57)
[...]
```

Please let me know if you consider this a bug which should be fixed or is it an expected behavior that was decided on which I might not be aware of. In the latter case maybe some note in documentation would be helpful as this kind of constrains the usage of JAX-RS in an application.

PS. there is also another issue that I've noticed in `OzarkContainerInitializer`. The classes provided to the `onStartup` are a set and as the iteration order cannot be relied upon and in fact with each restart/redeploy of the application the order is different, the observed behavior implemented by the class is different - the exception noted above is not thrown every time. Sometimes it is and sometimes the loop ends before the exception is thrown due to the `break` on line 67 (1.0.0-m03-SNAPSHOT).

**Comment by chkal** [comment]
Hey @tompasik,

thanks a lot for your feedback and sorry for the delayed response.

To be honest, I always thought that multiple Application classes are explicitly allowed by the JAX-RS spec, but I just checked the document and don't find any reference to that fact. That's really interesting. However, I think that people assume that this works with JAX-RS and it is supported by Jersey and other implementations.

Regarding Ozark: Actually it would be nice to support multiple application classes. But I'm not sure if this is easy to implement. After thinking about this I realized, that there are some application scoped components in Ozark, which could be a problem with multiple JAX-RS applications in one WAR. But I guess we have to dig deeper into this.

Thanks for bringing this up!

Regarding the other issue you mentioned. Could you perhaps create a separate issue for that or (even better) provide a pull request? Thanks!

**Comment by chkal** [comment]
This issue should be fixed with 511a4da4692f64fba7a311a7d7a21a39ff51bb96 in Ozark.

@tompasik It would be great if you could confirm the fix.

I don't think that we need to mention the support for multiple JAX-RS applications in a single web application explicitly in the spec. If anyone thinks this is necessary, feel free to open a corresponding issue in spec issue tracker.
